cita-chain/core/src/filters/rpc_filter.rs

```diff
@@ -0,0 +1,151 @@
+    fn get_filter_block(&self, id: usize) -> BlockNumber {
+        let filterdb = self.filter_db();
+        let mut block_filter = BlockNumber::min_value();
+
+        if let Some(block_number) = filterdb.try_lock().unwrap().get_block_filter(id) {
+            block_filter = *block_number;
+        }
+
+        drop(filterdb);
+        block_filter
+    }
+
+    fn get_logs_with_filter(&self, filter: Filter, block_filter: BlockNumber) -> Vec<Log> {
+        let mut filter: EthcoreFilter = filter.into();
+        filter.from_block = BlockTag::Height(block_filter);
+        filter.to_block = BlockTag::Tag(Tag::Latest);
+        let limit = filter.limit;
+        split_logs(
+            self.get_logs(&filter).into_iter().map(Into::into).collect(),
+            limit,
+        )
+    }
+}
+
+impl RpcFilter for Chain {
+    fn new_filter(&self, filter: Filter) -> usize {
+        let filterdb = self.filter_db();
+        let id = filterdb.try_lock().unwrap().gen_id();
+        filterdb.try_lock().unwrap().new_logs_filter(id, filter);
+        drop(filterdb);
+        id
+    }
+
+    fn new_block_filter(&self) -> usize {
+        let filterdb = self.filter_db();
+        let block_number = self.get_current_height();
+        let id = filterdb.try_lock().unwrap().gen_id();
+        filterdb
+            .try_lock()
+            .unwrap()
+            .new_block_filter(id, block_number);
+        drop(filterdb);
+        id
+    }
+
+    fn get_filter_changes(&self, id: usize) -> Option<FilterChanges> {
+        let filterdb = self.filter_db();
+        let mut changes = Some(FilterChanges::Empty);
+        let current_number = self.get_current_height();
+        let block_filter = self.get_filter_block(id);
+
+        if !filterdb.try_lock().unwrap().is_filter(id) {
+            drop(filterdb);
+            return changes;
+        }
+
+        // Check the logs
+        if let Some(filter) = filterdb.try_lock().unwrap().get_logs_filter(id) {
+            trace!("Into filter changes: logs");
+            changes = Some(FilterChanges::Logs(
+                self.get_logs_with_filter(filter.clone(), block_filter),
+            ));
+        };
+
+        // Check the block
+        if filterdb.try_lock().unwrap().is_block(id) {
+            trace!("Into filter changes: block");
+            // Return hashes from next block after the filter was created to the current block.
+            let hashes = ((block_filter + 1)..=current_number)
+                .filter_map(|_id| self.block_hash_by_height(_id))
+                .collect::<Vec<H256>>();
+            trace!("Block filter changes: {:?}", hashes);
+            changes = Some(FilterChanges::Hashes(
+                hashes.into_iter().map(Into::into).collect(),
+            ));
+        };
+
+        // Update the block filter
+        filterdb
+            .try_lock()
+            .unwrap()
+            .new_block_filter(id, current_number + 1);
+        drop(filterdb);
+        changes
+    }
```

Comment:
Why new such a block filter?

Reply:
Check the comment.
It's a bad name, i will refactor all the names later.
It's used to record the current block when call the `getFIlterChanges`.